src/main/scala/uk/gov/nationalarchives/tdr/api/utils/TreeNodesUtils.scala

```diff
@@ -0,0 +1,37 @@
+package uk.gov.nationalarchives.tdr.api.utils
+
+import java.io.{File => JIOFile}
+import java.util.UUID
+import uk.gov.nationalarchives.tdr.api.model.file.NodeType.{fileTypeIdentifier, folderTypeIdentifier}
+import uk.gov.nationalarchives.tdr.api.service.UUIDSource
+import uk.gov.nationalarchives.tdr.api.utils.TreeNodesUtils.TreeNode
+
+import scala.annotation.tailrec
+
+class TreeNodesUtils(uuidSource: UUIDSource) {
+  def generateNodes(filePaths: Set[String]): Map[String, TreeNode] = {
+    @tailrec
+    def innerFunction(originalPath: String, fileType: String, nodes: Map[String, TreeNode]): Map[String, TreeNode] = {
+      val pathWithoutInitialSlash: String = if (originalPath.startsWith("/")) originalPath.tail else originalPath
+      val jioFile = new JIOFile(pathWithoutInitialSlash)
+      val parentPath = Option(jioFile.getParent)
+      val name = jioFile.getName
+      val treeNode = TreeNode(uuidSource.uuid, name, parentPath, fileType)
+      val nextMap = nodes + (pathWithoutInitialSlash -> treeNode)
```

Comment:
Maybe this could just be done on line 29 instead of doing it each time `innerFunction` is called?

Reply:
I'm not sure what you mean. `nextMap` is recreated each time inner function is called with the current path and then returned at the end, I don't know how it can be moved out of `innerFunction`

Reply:
Oops, sorry; I meant to make this comment on line 15.

Reply:
Yeah, you're right, it only needs to be done once. I've moved it out of the `innerFunction` thanks.